lib/src/rocket.rs

```diff
@@ -191,17 +191,33 @@ impl Rocket {
         }
     }
 
-    #[inline]
+    #[inline(always)]
     pub(crate) fn dispatch<'s, 'r>(
+        &'s self,
+        request: &'r mut Request<'s>,
+        data: Data
+    ) -> Response<'r> {
+        //Dispatch the request and run all fairings on them
+        self.dispatch_impl(request, data, true)
+    }
+
+    #[inline]
+    fn dispatch_impl<'s, 'r>(
         &'s self,
         request: &'r mut Request<'s>,
         data: Data,
+        run_fairings: bool
     ) -> Response<'r> {
         info!("{}:", request);
 
         // Do a bit of preprocessing before routing; run the attached fairings.
         self.preprocess_request(request, &data);
-        self.fairings.handle_request(request, &data);
+
+        if run_fairings {
+            self.fairings.handle_request(request, &data);
+        }
+
+        let was_head_request = request.method() == Method::Head;
 
         // Route the request to get a response.
         let mut response = match self.route(request, data) {
```

Comment:
I think the better refactor is to split this big `match` out into another method, put everything before and after it in `dispatch`, and call the new method in the autohandled case instead of dispatch. In psuedocode:

```rust
fn dispatch(request, data) -> Response<'r> {
    // Remember whether this was originally a HEAD request or not.
    let head_request = request.method().is_head();

    // Preprocess the request and run the request fairings.
    self.preprocess_request(request, &data);
    self.fairings.handle_request(request, &data);

    // Route the request and run the user's handlers.
    let mut response = self.route_and_process(request, &data);

    // Run response fairings and strip the body if this was a HEAD request.
    self.fairings.handle_response(request, &mut response);
    if head_request { response.strip_body(); }

    response
}

fn route_and_process(request, data) -> Response<'r> {
    match self.route(request, data) {
        ...
        Outcome::Forward(data) => {
            if request.method() == Method::Head {
                request.set_method(Method::Get);
                self.route_and_process(request, data)
            } else {
                self.handle_error(Status::NotFound, request)
            }
        }
        ...
    }
}
```

This removes all of this `if` statements, clarifies the logic, and separates concerns nicely.

Reply:
Both this PR and the existing implementation were calling `self.preprocess_request()` twice as well. The refactor I propose makes it clear to see that this doesn't happen.

Reply:
@SergioBenitez thanks thats the best way we can do it. I will implement it next week and update the PR.